packages/dds/tree/docs/wip/schema-evolution.md

```diff
@@ -0,0 +1,583 @@
+
+   > In some environments, there's no guarantee that document saturation will ever occur - and this step is never completed.
+     Old documents can hang around forever if they remain unopened, but must still be supported in the event that someday they are needed.
+     The code that reads view schema A must be maintained forever.
+     Fortunately, this headache can be somewhat mitigated via the use of [adapters](#adapters).
+
+5. Eventually, all clients will have updated to the latest deployment, and all traces of schema A will be gone in both the documents and application source code.
+
+```mermaid
+gantt
+dateFormat D
+axisFormat Step %e
+tickInterval 1day
+Maintain view schema A in source code (read/write client) : 1, 2d
+Maintain view schema A in source code (read-only client)  : 1, 3d
+Documents contain stored schema A                         : 1, 3d
+Maintain view schema B in source code                     : 2, 3d
+Documents contain stored schema B                         : 3, 2d
+```
+
+Different applications with different requirements may tweak this process in various ways, but in general doing a staged rollout is a cumbersome endeavor.
+It takes time, it requires code bifurcation/duplication, and it's complicated to understand and coordinate.
+The next sections will explore ways to reduce the amount of effort a developer must take to change their schema.
+
+### Adapters
+
+A **schema adapter** helps to reduce the code bifurcation/duplication (described in the previous section) that results from doing a staged rollout.
+One of the main pain points of the staged rollout process is that the application has to maintain two view schema while the migration is ongoing (or forever!), since the client might encounter documents of either the old stored schema or the new stored schema.
+In practice, this can mean a lot of redundant application code as the UI is bound to two different data models, but only one is ever used at a time.
+Allowing legacy code to be removed earlier improves codebase health, aids developer velocity and reduces the size of shipped code.
+
+To facilitate this, an application can write a schema adapter.
+The adapter takes a tree of data that conforms to one schema and transforms it to conform to another schema.
+In the context of a schema migration, this means that the application can write code (e.g. the code that binds the view schema to the UI) _only against the new view schema_.
+The code that binds to the old view schema is deleted at the same time that the new view schema and the new adapter are added.
+This is a net benefit because it's simpler to write the adapter's code than it is to write the UI binding code.
+During the migration, when the application loads a document with the old stored schema, it first adapts the data in the document to the new view schema before the data is read by the app.
```

Comment:
I assume (from the steps 1-5 below) that adapters are bi-directional. I.e. they can convert A -> B, and as changes are made (in schema B), appropriate transformation is done to reflect changes in schema A.
If so, it might be worth mentioning it here